### PR TITLE
Feat: JSAP, Config, Input dir option, Output modes, Git patches generation

### PR DESCRIPTION
**Pull request by henry-lp**
@monperrus  @fermadeiral 

Change Logs:
* JSAP options
* Abstract legacy main for branching to new main 
* Can now specify workspace of outputs
* [Flyweight pattern] SonarQubeConfig
* [Strategy pattern] repair modes 
* Output modes - by default only output files modified by processor 
* Generate Git patches, if Git project repo root dir specified

Switch to new mode by specifying `--versionMode NEW` in main . By the default, it's the legacy mode running.

I am in a dilemma regarding Main implemeting MainApi due to legacy test cases make static calls to Main (`repair` and `normalrepair` method), since we would have to change the original test cases while the idea is to make zero changes to original code.

### CODE REVIEW

**Comment by fermadeiral** [comment]
@henry-lp, you are going to have conflict here when my PR is merged. Could you put this PR on hold until the release of the project, which will happen this week, so we can discuss about what you are doing, please? For the moment we just need a good ground, i.e., a clean repository, to release it, and then we can start to work on the functionality itself. Is that ok for you?

**Comment by henry-lp** [comment]
@fermadeiral yes, sounds good :) . I'll continue with playing with spoon and building the processors locally and make changes to repairnator to integrate this later.

**Comment by monperrus** [comment]
#63 merged, ready for rebase here!

**Comment by fermadeiral** [comment]
> @fermadeiral yes, sounds good :) . I'll continue with playing with spoon and building the processors locally and make changes to repairnator to integrate this later.

Perfect, Henry! I hope you are going well with Spoon!

**Comment by henry-lp** [comment]
@fermadeiral yes it tooks time going through the tutorial, docs and reading the papers. But I think I understand the basic to start experimenting :) . The links you sent me helped a lot.

**Comment by henry-lp** [comment]
@monperrus @fermadeiral  Green, ready for review :) .

**Comment by fermadeiral** [comment]
> @fermadeiral Green, ready for review :) .

I thought you were ok to wait for the release, but anyway...

You mentioned the changes, but I don't understand why you did them. What is the goal of this PR? Why are there different "main"s?

PS: if you are using IntelliJ, please, use "Reformat Code" or something equivalent in the IDE you are using, the indentation here is broken.

**Comment by henry-lp** [comment]
@fermadeiral 

Right this can wait. I forgot, It's a habit of my to request review after CI got green :) . 

No I don't use Intellj only Sublime text editor. The point is to abstracting the Main to perserve to legacy functionalities until we trust the new Main, then you can just delete the old Main and move the new Main out of its folder :) . 

I want this structure since we can have different repair modes later, it's nice to have centralized config and instead of raw input extract we can use JSAP.

**Comment by monperrus** [comment]
conflict?

**Comment by henry-lp** [comment]
> 
> 
> conflict?

yes, It's resolved :) .

**Comment by monperrus** [comment]
Thanks @henry-lp! @fermadeiral you have merge rights. would you want to handle merging here? Thanks!

**Comment by henry-lp** [comment]
@fermadeiral changed per discussion :) . Ready for review .

**Comment by henry-lp** [comment]
@fermadeiral all fixed :) . Review ready again.

**Comment by henry-lp** [comment]
new feature: only files changed will be output as patches in latest commit :) .

**Comment by monperrus** [comment]
@fermadeiral I assume you're responsible for merging Henry's PR, OK for you?

**Comment by henry-lp** [comment]
I will update `usage` in README once the PR is done review and merge :) .

File: pom.xml [hunk 1]
```diff
@@ -15,56 +12,109 @@
...
+                        <id>default</id>
+                        <goals>
+                            <goal>perform</goal>
+                        </goals>
+                        <configuration>
+                            <pomFileName>pom.xml</pomFileName>
+                        </configuration>
+                    </execution>
+                </executions>
+            </plugin>
```
Comment (fermadeiral) [inline pom.xml:117]: There is a duplicate block of code here.

File: src/main/java/sonarquberepair/DefaultRepair.java [hunk 2]
```diff
@@ -0,0 +1,96 @@
+package sonarquberepair;
+
+import spoon.Launcher;
+import spoon.processing.Processor;
+import spoon.support.sniper.SniperJavaPrettyPrinter;
+
+import java.lang.reflect.Constructor;
+import java.io.File;
+import java.util.List;
+import java.nio.file.Paths;
```
Comment (fermadeiral) [inline src/main/java/sonarquberepair/DefaultRepair.java:10]: Unused import.

File: src/main/java/sonarquberepair/DefaultRepair.java [hunk 3]
```diff
@@ -0,0 +1,96 @@
...
+
+
+import sonarquberepair.Processors;
```
Comment (fermadeiral) [inline src/main/java/sonarquberepair/DefaultRepair.java:13]: Unused import.

File: src/main/java/sonarquberepair/DefaultRepair.java [hunk 4]
```diff
@@ -0,0 +1,96 @@
...
+import spoon.support.JavaOutputProcessor;
+import spoon.support.QueueProcessingManager;
+import spoon.processing.ProcessingManager;
+import spoon.reflect.factory.Factory;
+import spoon.reflect.CtModel;
+import spoon.IncrementalLauncher;
```
Comment (fermadeiral) [inline src/main/java/sonarquberepair/DefaultRepair.java:19]: Unused import.

File: src/main/java/sonarquberepair/FileOutputStrategy.java [hunk 5]
```diff
@@ -0,0 +1,6 @@
+package sonarquberepair;
+
+public enum FileOutputStrategy {
+	ALL,
+	ONLYCHANGED
```
Comment (fermadeiral) [inline src/main/java/sonarquberepair/FileOutputStrategy.java:5]: Replace by `ONLY_CHANGED`.

File: src/main/java/sonarquberepair/Main.java [hunk 6]
```diff
@@ -1,66 +1,106 @@
 package sonarquberepair;
 
-import spoon.Launcher;
-import spoon.processing.Processor;
 import spoon.support.sniper.SniperJavaPrettyPrinter;
```
Comment (fermadeiral) [inline src/main/java/sonarquberepair/Main.java:3]: Unused import.

File: src/main/java/sonarquberepair/Main.java [hunk 7]
```diff
@@ -1,66 +1,106 @@
...
+
+	public SonarQubeRepairConfig getConfig() {
+		return this.config;
+	}
+
+	public JSAP defineArgs() throws JSAPException{
+		JSAP jsap = new JSAP();
+
+		/* will be supporting multiple rules processing later so rulenumber(s) */
+		FlaggedOption opt = new FlaggedOption("ruleNumbers");
```
Comment (fermadeiral) [inline src/main/java/sonarquberepair/Main.java:22]: `ruleNumbers` should be replaced by `ruleKeys` (check other places for similar situations and change them). We should keep out terminology as closest as possible to SonarSource.

File: src/main/java/sonarquberepair/Main.java [hunk 8]
```diff
@@ -1,66 +1,106 @@
...
+		opt.setLongFlag("workspace");
+		opt.setStringParser(JSAP.STRING_PARSER);
+		opt.setDefault("./sonar-workspace");
+		opt.setHelp("Workspace of SonarQubeRepair");
+		jsap.registerParameter(opt);
+
+		opt = new FlaggedOption("gitRepoPath");
+		opt.setLongFlag("gitRepoPath");
+		opt.setStringParser(JSAP.STRING_PARSER);
+		opt.setHelp("Workspace of SonarQubeRepair");
```
Comment (fermadeiral) [inline src/main/java/sonarquberepair/Main.java:67]: This description is a bad copy/paste.

File: src/main/java/sonarquberepair/Main.java [hunk 9]
```diff
@@ -1,66 +1,106 @@
...
+		this.getConfig().setRepairPath(jsapRes.getString("repairPath"));
+		this.getConfig().setPrettyPrintingStrategy(PrettyPrintingStrategy.valueOf(jsapRes.getString("prettyPrintingStrategy")));
+		this.getConfig().setFileOutputStrategy(FileOutputStrategy.valueOf(jsapRes.getString("fileOutputStrategy")));
+		this.getConfig().setWorkSpace(jsapRes.getString("workspace"));
+		this.getConfig().setGitRepoPath(jsapRes.getString("gitRepoPath"));
+	}
+
+
+	public DefaultRepair getRepairProcess() {
+		DefaultRepair repairProc = new DefaultRepair(this.config);
```
Comment (fermadeiral) [inline src/main/java/sonarquberepair/Main.java:87]: Replace `repairProc` by `defaultRepair`.

File: src/main/java/sonarquberepair/Main.java [hunk 10]
```diff
@@ -1,66 +1,106 @@
...
-			projectKey = args[1];
-		}
-		System.out.println("Applying " + Processors.getProcessor(ruleKey).getName() + "...");
-		repair("./source/act/", projectKey, ruleKey, PrettyPrintingStrategy.SNIPER);
-		System.out.println("Done.");
+	public void start(String[] args) throws Exception {
+		JSAP jsap = defineArgs();
+		this.initConfig(jsap,args);
+		this.getRepairProcess().repair();
+		System.out.println("done");
```
Comment (fermadeiral) [inline src/main/java/sonarquberepair/Main.java:99]: Maybe those lines of code could be directly in `main(String[] args)`, WDYT?

File: src/main/java/sonarquberepair/SonarQubeRepairConfig.java [hunk 11]
```diff
@@ -0,0 +1,73 @@
+package sonarquberepair;
+
+import java.util.List;
+import java.util.ArrayList;
+
+/* All config settings of SonarQube should be gathered here */
+public class SonarQubeRepairConfig {
+	private final List<Integer> ruleNumbers = new ArrayList<Integer>();
```
Comment (fermadeiral) [inline src/main/java/sonarquberepair/SonarQubeRepairConfig.java:8]: Replace `ruleNumbers` by `ruleKeys` (and fix methods about it accordingly).

File: src/main/java/sonarquberepair/UniqueTypesCollector.java [hunk 12]
```diff
@@ -0,0 +1,42 @@
...
+import java.util.HashMap;
+import java.util.Map;
+import spoon.reflect.declaration.CtType;
+import spoon.reflect.declaration.CtElement;
+
+/* Only add the CtType object if it does not exist in the map yet */
+public class UniqueTypesCollector {
+	private static UniqueTypesCollector uniqueTypesCollector;
+
+	private Map<String,CtType> topLevelTypes4Output = new HashMap<String,CtType>();
```
Comment (fermadeiral) [inline src/main/java/sonarquberepair/UniqueTypesCollector.java:12]: I believe it's better to do such initialization in the constructor.

File: src/main/java/sonarquberepair/SonarQubeRepairConfig.java [hunk 13]
```diff
@@ -0,0 +1,73 @@
...
+
+	public void setRepairPath(String repairPath) {
+		this.repairPath = repairPath;
+	}
+
+	public String getRepairPath() {
+		return this.repairPath;
+	}
+
+	public void setWorkSpace(String workspace) {
```
Comment (fermadeiral) [inline src/main/java/sonarquberepair/SonarQubeRepairConfig.java:58]: Replace `setWorkSpace` by `setWorkspace`.

File: src/main/java/sonarquberepair/SonarQubeRepairConfig.java [hunk 14]
```diff
@@ -0,0 +1,73 @@
...
+		this.workspace = workspace;
+	}
+
+	public String getWorkSpace() {
```
Comment (fermadeiral) [inline src/main/java/sonarquberepair/SonarQubeRepairConfig.java:62]: Replace `getWorkSpace` by `getWorkspace`.

File: src/main/java/sonarquberepair/SonarQubeRepairConfig.java [hunk 15]
```diff
@@ -0,0 +1,73 @@
...
+import java.util.List;
+import java.util.ArrayList;
+
+/* All config settings of SonarQube should be gathered here */
+public class SonarQubeRepairConfig {
+	private final List<Integer> ruleNumbers = new ArrayList<Integer>();
+	private String projectKey;
+	private PrettyPrintingStrategy prettyPrintingStrategy;
+	private FileOutputStrategy fileOutputStrategy;
+	private String repairPath;
```
Comment (fermadeiral) [inline src/main/java/sonarquberepair/SonarQubeRepairConfig.java:12]: `repairPath` is not an intuitive name. What about `originalFilesPath`?

**Comment by henry-lp** [comment]
Ready to merge when green :) .

File: ChangedOnlyFilesOutput/SonarGitPatches/sonarpatch_1.patch [hunk 16]
```diff
@@ -0,0 +1,69 @@
+--- a/src/test/resources/BigDecimalDoubleConstructor.java
```
Comment (fermadeiral) [inline ChangedOnlyFilesOutput/SonarGitPatches/sonarpatch_1.patch:1]: Is this file supposed to be in this PR?

File: src/main/java/sonarquberepair/Main.java (hunk 6 above)
Comment (fermadeiral) [inline src/main/java/sonarquberepair/Main.java:3]: Unused import.

File: src/main/java/sonarquberepair/Main.java [hunk 17]
```diff
@@ -1,66 +1,106 @@
...
+
+	public JSAP defineArgs() throws JSAPException{
+		JSAP jsap = new JSAP();
+
+		/* will be supporting multiple rules processing later so rulenumber(s) */
+		FlaggedOption opt = new FlaggedOption("ruleKeys");
+		opt.setLongFlag("ruleKeys");
+		opt.setStringParser(JSAP.INTEGER_PARSER);
+		opt.setDefault("2116");
+		opt.setHelp("Sonarqube rule number, Check https://rules.sonarsource.com/java");
```
Comment (fermadeiral) [inline src/main/java/sonarquberepair/Main.java:26]: "rule number" -> "rule key".

Always be consistent.

File: src/main/java/sonarquberepair/Main.java [hunk 18]
```diff
@@ -1,66 +1,106 @@
...
+		opt.setStringParser(JSAP.STRING_PARSER);
+		opt.setDefault("fr.inria.gforge.spoon:spoon-core");
+		opt.setHelp("what is this projectKey ?");
+		jsap.registerParameter(opt);
+
+		opt = new FlaggedOption("originalFilesPath");
+		opt.setLongFlag("originalFilesPath");
+		opt.setStringParser(JSAP.STRING_PARSER);
+		opt.setDefault("./source/act/");
+		opt.setHelp("The input folder or file for SonarQube Repair to work on");
```
Comment (fermadeiral) [inline src/main/java/sonarquberepair/Main.java:40]: "SonarQube Repair" -> "sonarqube-repair"

File: src/main/java/sonarquberepair/Main.java [hunk 19]
```diff
@@ -1,66 +1,106 @@
...
-			cons = processor.getConstructor();
-			object = cons.newInstance();
-		}
-		launcher.addProcessor((Processor) object);
-		launcher.run();
+		opt = new FlaggedOption("workspace");
+		opt.setLongFlag("workspace");
+		opt.setStringParser(JSAP.STRING_PARSER);
+		opt.setDefault("./sonar-workspace");
+		opt.setHelp("Workspace of SonarQubeRepair");
```
Comment (fermadeiral) [inline src/main/java/sonarquberepair/Main.java:61]: "SonarQubeRepair" -> "sonarqube-repair"

File: src/main/java/sonarquberepair/Main.java [hunk 20]
```diff
@@ -1,66 +1,106 @@
...
+		jsap.registerParameter(opt);
+
+		opt = new FlaggedOption("gitRepoPath");
+		opt.setLongFlag("gitRepoPath");
+		opt.setStringParser(JSAP.STRING_PARSER);
+		opt.setHelp("Root Path of the Git repo directory");
```
Comment (fermadeiral) [inline src/main/java/sonarquberepair/Main.java:67]: Which git repo directory?

File: src/main/java/sonarquberepair/Main.java [hunk 21]
```diff
@@ -1,66 +1,106 @@
...
+		return defaultRepair;
 	}
 
 	/**
-	 * @param args string array.
-	 *             Give one or two arguments.
-	 *             The first argument is the Sonar rule key.
-	 *             The second argument is the project key for the Sonar analysis of source files.
+	 * @param args string array. Give either 0, 1 or 2 arguments. first argument is sonarqube rule-number which you can get from https://rules.sonarsource.com/java/type/Bug
+	 *             second argument is the projectKey for the sonarqube analysis of source files. for  example "fr.inria.gforge.spoon:spoon-core"
```
Comment (fermadeiral) [inline src/main/java/sonarquberepair/Main.java:93]: Henry, you should undo this change in the comment. I believe you didn't do this change on purpose, it happened in some reset you did or something like that.

File: src/main/java/sonarquberepair/Main.java (hunk 10 above)
Comment (fermadeiral) [reply src/main/java/sonarquberepair/Main.java:99]: WDYT?

File: src/main/java/sonarquberepair/SonarQubeRepairConfig.java [hunk 22]
```diff
@@ -0,0 +1,73 @@
...
+
+	public void setProjectKey(String projectKey) {
+		this.projectKey = projectKey;
+	}
+
+	public String getProjectKey() {
+		return this.projectKey;
+	}
+
+	public void addRuleKeys(int ruleNumber) {
```
Comment (fermadeiral) [inline src/main/java/sonarquberepair/SonarQubeRepairConfig.java:26]: `ruleNumber` -> `ruleKey`

File: src/test/java/sonarquberepair/processor/spoonbased/CastArithmeticOperandTest.java [hunk 23]
```diff
@@ -14,10 +14,16 @@
     public void test() throws Exception {
         String fileName = "CastArithmeticOperand.java";
         String pathToBuggyFile = Constants.PATH_TO_FILE + fileName;
-        String pathToRepairedFile = "./spooned/" + fileName;
+        String pathToRepairedFile = Constants.WORKSPACE + "/spooned/" + fileName;
 
+        System.out.println(pathToBuggyFile + " " + pathToRepairedFile);
```
Comment (fermadeiral) [inline src/test/java/sonarquberepair/processor/spoonbased/CastArithmeticOperandTest.java:19]: Why does this line exist?

File: src/main/java/sonarquberepair/Main.java (hunk 20 above)
Comment (henry-lp) [reply src/main/java/sonarquberepair/Main.java:67]: The originalFilesPath might not be the same as this. We might have multiple other subdir and only one of them is the Git repo dir which we need to specify where it is to generate a non-corrupt patch.

File: src/main/java/sonarquberepair/Main.java (hunk 21 above)
Comment (henry-lp) [reply src/main/java/sonarquberepair/Main.java:93]: removed again :) .

File: src/test/java/sonarquberepair/processor/spoonbased/CastArithmeticOperandTest.java (hunk 23 above)
Comment (henry-lp) [reply src/test/java/sonarquberepair/processor/spoonbased/CastArithmeticOperandTest.java:19]: Debug message from before. I have removed it :) .

**Comment by henry-lp** [comment]
>>> WDYT?

I think we should keep it that way otherwise I would have to change modifier of all other methods in start to `public static` in main method , which I don't like very much :) . WDYT ?

**Comment by fermadeiral** [comment]
> I think we should keep it that way otherwise I would have to change modifier of all other methods in start to public static in main method , which I don't like very much :) . WDYT ?

I think you are right to leave it that way!

File: src/main/java/sonarquberepair/SonarQubeRepairConfig.java [hunk 24]
```diff
@@ -23,8 +23,8 @@ public String getProjectKey() {
 		return this.projectKey;
 	}
 
-	public void addRuleKeys(int ruleNumber) {
-		this.ruleKeys.add(ruleNumber);
+	public void addRuleKeys(int rulekey) {
```
Comment (fermadeiral) [inline src/main/java/sonarquberepair/SonarQubeRepairConfig.java:26]: Always use camel case. Almost done!!!

Comment (henry-lp) [reply src/main/java/sonarquberepair/SonarQubeRepairConfig.java]: done :)

**Comment by henry-lp** [comment]
Thanks for the extreme detail review, Fernanda :) .

**Comment by fermadeiral** [comment]
Thank you, Henry! Good job, and congrats on your first PR in this repo!

**Comment by monperrus** [comment]
Thanks a lot to both of you!
